storm-core/src/jvm/org/apache/storm/metrics2/StormMetricRegistry.java

```diff
@@ -0,0 +1,181 @@
+        String metricName = metricName(name, topologyId, componentId, streamId,taskId, workerPort);
+        return REGISTRY.counter(metricName);
+    }
+
+    public static void start(Map<String, Object> stormConfig, DaemonType type){
+        try {
+            hostName = dotToUnderScore(Utils.localHostname());
+        } catch (UnknownHostException e) {
+             LOG.warn("Unable to determine hostname while starting the metrics system. Hostname will be reported" +
+                     " as 'localhost'.");
+        }
+
+        LOG.info("Starting metrics reporters...");
+        List<Map<String, Object>> reporterList = (List<Map<String, Object>>)stormConfig.get(Config.STORM_METRICS_REPORTERS);
+        if(reporterList != null && reporterList.size() > 0) {
+            for (Map<String, Object> reporterConfig : reporterList) {
+                // only start those requested
+                List<String> daemons = (List<String>) reporterConfig.get("daemons");
+                for (String daemon : daemons) {
+                    if (DaemonType.valueOf(daemon.toUpperCase()) == type) {
+                        startReporter(stormConfig, reporterConfig);
+                    }
+                }
+            }
+        }
+    }
+
+    public static MetricRegistry registry(){
+        return REGISTRY;
+    }
+
+    private static void startReporter(Map<String, Object> stormConfig, Map<String, Object> reporterConfig){
+        String clazz = (String)reporterConfig.get("class");
+        StormReporter reporter = null;
+        LOG.info("Attempting to instantiate reporter class: {}", clazz);
+        reporter = Utils.newInstance(clazz);
+        if(reporter != null){
+            reporter.prepare(REGISTRY, stormConfig, reporterConfig);
+            reporter.start();
+            REPORTERS.add(reporter);
+        }
+
+    }
+
+    public static void stop(){
+        for(StormReporter sr : REPORTERS){
+            sr.stop();
+        }
+    }
+
+    public static String metricName(String name, String stormId, String componentId, String streamId, Integer taskId, Integer workerPort){
+        StringBuilder sb = new StringBuilder("storm.worker.");
+        sb.append(stormId);
+        sb.append(".");
+        sb.append(hostName);
+        sb.append(".");
+        sb.append(dotToUnderScore(componentId));
+        sb.append(".");
+        sb.append(dotToUnderScore(streamId));
+        sb.append(".");
+        sb.append(taskId);
+        sb.append(".");
+        sb.append(workerPort);
+        sb.append("-");
+        sb.append(name);
+        return sb.toString();
+    }
+
+    public static String metricName(String name, String stormId, String componentId, Integer workerPort) {
+        StringBuilder sb = new StringBuilder("storm.worker.");
+        sb.append(stormId);
+        sb.append(".");
+        sb.append(hostName);
+        sb.append(".");
+        sb.append(dotToUnderScore(componentId));
```

Comment:
why is there no task id?  Meters and Counters get a taskId, but Guages do not?  I assume that is just how these metrics are being used right now.  I personally would rather have a task id than a component id, because I can recreate the component id.

Reply:
Gauges (currently) are specific to disruptor metrics which are contextually different from some of the other metrics. Let me know what data you'd like to see in the paths/names for the disruptor metrics, and I'd be happy to add them.

Reply:
I would like to see a task id simply to distinguish between different executors within the same worker.  Perhaps we could use the smallest task id within the executor for these metrics.

Reply:
What about the case of `worker-transfer-queue`?

Reply:
The worker transfer queue should have a -1 task id, and the component id should be __system, just like the system bolt.  This way it is consistent for worker level metrics.

Reply:
Fixed in latest commit.